libgraph/src/GraphTopology.cpp

```diff
@@ -422,6 +422,118 @@ katana::EdgeShuffleTopology::SortEdgesByTypeThenDest(
+        auto e_end = *OutEdges(node).end();
+
+        // Get iterators to locations to sort in the vector
+        auto begin_sort_iter = katana::make_zip_iterator(
+            edge_prop_indices_.begin() + e_beg, GetDests().begin() + e_beg);
+        auto end_sort_iter = katana::make_zip_iterator(
+            edge_prop_indices_.begin() + e_end, GetDests().begin() + e_end);
+
+        // Sort a graph topology by specified edge property and then by
+        // edge destinations.
+        // Rearrange property vector indices based on how this graph will
+        // eventually be sorted.
+        std::sort(
+            begin_sort_iter, end_sort_iter,
+            [&](const auto& tup1, const auto& tup2) {
+              // Get edge property values and destinations
+              auto e1 = std::get<0>(tup1);
+              auto e2 = std::get<0>(tup2);
+              static_assert(
+                  std::is_same_v<
+                      decltype(e1), katana::GraphTopology::PropertyIndex>);
+              static_assert(
+                  std::is_same_v<
+                      decltype(e2), katana::GraphTopology::PropertyIndex>);
+
+              auto e1_prop_idx = pg->GetEdgePropertyIndexFromOutEdge(e1);
+              auto e2_prop_idx = pg->GetEdgePropertyIndexFromOutEdge(e2);
+
+              auto data1 = edge_prop_view->Value(e1_prop_idx);
+              auto data2 = edge_prop_view->Value(e2_prop_idx);
+              if (data1 != data2) {
+                return data1 < data2;
+              }
+
+              auto dst1 = std::get<1>(tup1);
+              auto dst2 = std::get<1>(tup2);
+              static_assert(
+                  std::is_same_v<decltype(dst1), katana::GraphTopology::Node>);
+              static_assert(
+                  std::is_same_v<decltype(dst2), katana::GraphTopology::Node>);
+              return dst1 < dst2;
+            });
+      },
+      katana::steal(), katana::no_stats());
+}
+
+void
+katana::EdgeShuffleTopology::SortEdgesByPropertyThenDest(
+    const PropertyGraph* pg,
+    const std::optional<std::string> property_name) noexcept {
+  if (!property_name) {
+    KATANA_LOG_FATAL(
+        "This function must need the target property name", "for sorting");
+  }
+  // Get the `ChunkedArray` corresponding to `property_name`.
+  auto edge_prop_ret = pg->GetEdgeProperty(property_name.value());
+  if (!edge_prop_ret) {
+    KATANA_LOG_FATAL(
+        "PropertyGraph does not have property: {}", property_name.value());
+  }
+  auto edge_property = edge_prop_ret.value();
+  // TODO(lhc): For now, this finds a property twice through two APIs,
+  //            by GetEdgeProperty() and GetEdgePropertyTyped().
+  //            Later, implement and merge those two functions, and hide
+  //            getting types from developers.
+  //            This is because first, its focus is not performances,
+  //            and second, want to use the existing APIs as much as possible.
+  switch (edge_property->type()->id()) {
```

Comment:
See `libsupport/include/ArrowVisitor.h`. This seems like a case where an ArrowArrayVisitor can help. You'll have to look up some example to learn how to use it.